src/Microsoft.Azure.SignalR.Common/ClientInvocation/ClientResultsManager.cs

```diff
@@ -0,0 +1,312 @@
+                // or someone else completed the invocation (likely a bad client)
+                // we'll ignore both cases
+                if (_pendingInvocations.Remove(message.InvocationId!, out _))
+                {
+                    item.Complete(item.Tcs, message);
+                }
+            }
+            else
+            {
+                // connection was disconnected or someone else completed the invocation
+            }
+        }
+
+        public bool CheckRoutedInvocation(string invocationId)
+        {
+            return _routedInvocations.TryGetValue(invocationId, out var _);
+        }
+
+        public void TryCompleteRoutedResult(string connectionId, CompletionMessage message)
+        {
+            if (_routedInvocations.TryGetValue(message.InvocationId!, out var item))
+            {
+                if (item.ConnectionId != connectionId)
+                {
+                    throw new InvalidOperationException($"Connection ID '{connectionId}' is not valid for invocation ID '{message.InvocationId}'.");
+                }
+
+                // if false the connection disconnected right after the above TryGetValue
+                // or someone else completed the invocation (likely a bad client)
+                // we'll ignore both cases
+                if (_routedInvocations.Remove(message.InvocationId!, out _))
+                {
+                    item.Complete(item.Tcs, message);
+                }
+            }
+            else
+            {
+                // connection was disconnected or someone else completed the invocation
+            }
+        }
+
+        public (Type Type, string ConnectionId, object Tcs, Action<object, CompletionMessage> Completion)? RemoveInvocation(string invocationId)
+        {
+            _pendingInvocations.TryRemove(invocationId, out var item);
+            return item;
+        }
+
+        public (Type Type, string ConnectionId, string callerServerId, object Tcs, Action<object, CompletionMessage> Completion)? RemoveRoutedInvocation(string invocationId)
+        {
+            _routedInvocations.TryRemove(invocationId, out var item);
+            return item;
+        }
+
+        public Type GetReturnType(string invocationId)
+        {
+            if (TryGetType(invocationId, out var type))
+            {
+                return type;
+            }
+            throw new InvalidOperationException($"Invocation ID '{invocationId}' is not associated with a pending client result.");
+        }
+
+        public bool TryGetType(string invocationId, out Type type)
+        {
+            if (_pendingInvocations.TryGetValue(invocationId, out var item1))
+            {
+                type = item1.Type;
+                return true;
+            }
+            if (_routedInvocations.TryGetValue(invocationId, out var item2))
+            {
+                type = item2.Type;
```

Comment:
For routed ones, it seems always return `object` refer to the add method. Simply make type null and return true?

Reply:
`type` is update to `RawResult`. See details in [this](https://github.com/Azure/azure-signalr/pull/1684#discussion_r980455654) and [this](https://github.com/Azure/azure-signalr/pull/1684#discussion_r982699179)